app/lib/search/index_ducene.dart

```diff
@@ -0,0 +1,226 @@
+      ..append('version', doc.version)
+      ..append('devVersion', doc.devVersion)
+      ..append(
+        'description',
+        _normalizeText(doc.description, 500),
+        analyzer: _analyzer,
+        stored: false,
+      )
+      ..append(
+          'originalDescription', _normalizeWhitespaces(doc.description, 500))
+      ..append('lastUpdated', doc.lastUpdated)
+      ..append(
+        'readme',
+        _normalizeText(doc.readme, 2000),
+        analyzer: _analyzer,
+        stored: false,
+      )
+      ..append('popularity', doc.popularity.toString());
+    if (doc.detectedTypes != null && doc.detectedTypes.isNotEmpty) {
+      duceneDoc.append('detectedTypes', doc.detectedTypes);
+    }
+    await _index.updateDocuments([duceneDoc]);
+  }
+
+  @override
+  Future addAll(Iterable<PackageDocument> documents) async {
+    // Adding documents one-by-one, will need to experiment with better batch
+    // size when async disk index is available.
+    for (PackageDocument doc in documents) {
+      await add(doc);
+    }
+  }
+
+  @override
+  Future removeUrl(String url) async {
+    await _init();
+    await _index.deleteDocuments(new BoolQuery()..append('id', url));
+  }
+
+  @override
+  Future merge() async {
+    await _init();
+    await _index.forceMerge();
+    _lastUpdated = new DateTime.now();
+  }
+
+  @override
+  Future<PackageSearchResult> search(PackageQuery query) async {
+    if (!isReady) return new PackageSearchResult.notReady();
+
+    await _init();
+    final IndexSearcher searcher = _index.newRealTimeIndexSearcher();
+    final String queryText = _normalizeText(query.text, 200);
+    final duceneQuery = new BoolQuery()
+      ..append('package', queryText, analyzer: _analyzer, boost: 4.0)
+      ..append('description', queryText, analyzer: _analyzer, boost: 2.0)
+      ..append('readme', queryText, analyzer: _analyzer);
+    if (query.type != null) {
+      duceneQuery
+          .addFilter(new TermQuery(new Term('detectedTypes', query.type)));
+    }
+
+    final int offset = query.offset ?? 0;
+    final int limit = max(minSearchLimit, query.limit ?? 100);
+
+    final TopDocs topPackages = await searcher.search(
+      duceneQuery,
+      maxSearchResults + minSearchLimit, // additional results for extended bias
+      scoreSort: new TFIDFScoreSort(),
+    );
+
+    List<PackageScore> packages = [];
+    for (ScoreDoc sd in topPackages.scoreDocs) {
+      final Document doc = await searcher.doc(sd.doc);
+      final double bias = double.parse(doc.get('popularity')) * 20;
+
+      packages.add(new PackageScore(
+        doc: new PackageDocument(
+          url: doc.get('id'),
+          package: doc.get('package'),
+          version: doc.get('version'),
+          devVersion: doc.get('devVersion'),
+          description: doc.get('originalDescription'),
+          lastUpdated: doc.get('lastUpdated'),
+          detectedTypes: doc.getValues('detectedTypes'),
+        ),
+        score: sd.score + bias,
```

Comment:
Maybe add a comment here how these two doubles get combined. What is the range of them? Do they have 50/50 weight?